File: apps/api/schemas/domain.py
"""
Pydantic schemas for domain entities.

These schemas define the request/response formats for the admin API endpoints.
"""

from datetime import datetime
from uuid import UUID
from typing import Optional

from pydantic import BaseModel, EmailStr, Field

from models.property import PropertyStatus
from models.florist import FloristStatus
from models.user import UserRole, SubscriptionStatus
from models.delivery import SubscriptionPlan


# =============================================================================
# Property Schemas
# =============================================================================

class PropertyCreate(BaseModel):
    """Schema for creating a new property."""
    name: str = Field(..., min_length=1, max_length=255, description="Property name")
    address: str = Field(..., min_length=1, max_length=500, description="Property address")
    delivery_cadence: Optional[str] = Field(
        None,
        max_length=100,
        description="Delivery schedule (e.g., 'weekly', 'bi-weekly')"
    )


class PropertyUpdate(BaseModel):
    """Schema for updating an existing property."""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    address: Optional[str] = Field(None, min_length=1, max_length=500)
    status: Optional[PropertyStatus] = Field(None, description="New status for the property")
    delivery_cadence: Optional[str] = Field(None, max_length=100)


class PropertyResponse(BaseModel):
    """Schema for property API responses."""
    id: UUID
    name: str
    address: str
    status: PropertyStatus
    delivery_cadence: Optional[str]
    created_at: datetime
    updated_at: datetime
    
    model_config = {"from_attributes": True}


class EnrichedPropertyResponse(BaseModel):
    """
    Enhanced property response with computed fields.
    
    Includes user counts, florist name, and property manager email
    for the admin properties table.
    """
    id: UUID
    name: str
    address: str
    status: PropertyStatus
    delivery_cadence: Optional[str]
    total_users: int = Field(default=0, description="Count of all users associated with this property")
    active_users: int = Field(default=0, description="Count of users with ACTIVE subscription status")
    florist_name: Optional[str] = Field(default=None, description="Name of assigned florist")
    property_manager_email: Optional[str] = Field(default=None, description="Email of assigned property manager")
    created_at: datetime
    updated_at: datetime
    
    model_config = {"from_attributes": True}


class AssignPMRequest(BaseModel):
    """Schema for assigning a property manager to a property."""
    user_id: UUID = Field(..., description="ID of the user to assign as property manager")


# =============================================================================
# Florist Schemas
# =============================================================================

class FloristCreate(BaseModel):
    """Schema for creating a new florist."""
    name: str = Field(..., min_length=1, max_length=255, description="Florist business name")


class FloristResponse(BaseModel):
    """Schema for florist API responses."""
    id: UUID
    name: str
    status: FloristStatus
    created_at: datetime
    
    model_config = {"from_attributes": True}


# =============================================================================
# PropertyAssignment Schemas
# =============================================================================

class PropertyAssignmentCreate(BaseModel):
    """Schema for creating a new property-florist assignment."""
    property_id: UUID = Field(..., description="ID of the property to assign")
    florist_id: UUID = Field(..., description="ID of the florist to assign")


class PropertyAssignmentResponse(BaseModel):
    """Schema for property assignment API responses."""
    id: UUID
    property_id: UUID
    florist_id: UUID
    active: bool
    created_at: datetime
    
    model_config = {"from_attributes": True}


# =============================================================================
# Registration Schemas (Public)
# =============================================================================

class RegisterRequest(BaseModel):
    """Schema for customer self-registration."""
    email: EmailStr = Field(..., description="User email address")
    password: str = Field(..., min_length=6, description="User password (min 6 characters)")


class RegisterResponse(BaseModel):
    """Schema for registration response (same as login)."""
    access_token: str
    token_type: str = "bearer"
    user: "UserResponseWithOnboarding"


class UserResponseWithOnboarding(BaseModel):
    """User response with onboarding fields for registration/login."""
    id: UUID
    email: EmailStr
    role: UserRole
    property_id: Optional[UUID] = None
    subscription_status: Optional[SubscriptionStatus] = None
    subscription_plan: Optional[SubscriptionPlan] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MeResponseWithPropertyName(BaseModel):
    """
    User response with property details for customer dashboard.

    Includes property_name and property_address resolved from property_id for display.
    """
    id: UUID
    email: EmailStr
    role: UserRole
    property_id: Optional[UUID] = None
    property_name: Optional[str] = Field(default=None, description="Resolved property name")
    property_address: Optional[str] = Field(default=None, description="Resolved property address")
    unit: Optional[str] = Field(default=None, description="User's unit number within the property")
    subscription_status: Optional[SubscriptionStatus] = None
    subscription_plan: Optional[SubscriptionPlan] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# =============================================================================
# User Schemas
# =============================================================================

class UserCreate(BaseModel):
    """Schema for creating a new user."""
    email: EmailStr = Field(..., description="User email address")
    role: UserRole = Field(..., description="User role")
    password: str = Field(..., min_length=6, description="User password (min 6 characters)")
    property_id: Optional[UUID] = Field(
        None,
        description="Property ID (only valid for PROPERTY_MANAGER role)"
    )


class UserUpdate(BaseModel):
    """Schema for updating an existing user."""
    role: Optional[UserRole] = Field(None, description="New role for the user")
    property_id: Optional[UUID] = Field(
        None,
        description="Property ID (only valid for PROPERTY_MANAGER role)"
    )
    subscription_status: Optional[SubscriptionStatus] = Field(
        None,
        description="Subscription status (only valid for CUSTOMER role)"
    )


class EnrichedUserResponse(BaseModel):
    """
    Enhanced user response with resolved property name.
    
    Includes property_name resolved from property_id for display in admin table.
    subscription_status is null for non-CUSTOMER users.
    """
    id: UUID
    email: EmailStr
    role: UserRole
    property_id: Optional[UUID] = None
    property_name: Optional[str] = Field(default=None, description="Resolved property name")
    subscription_status: Optional[SubscriptionStatus] = Field(
        default=None,
        description="Subscription status (null for non-CUSTOMER users)"
    )
    created_at: datetime
    
    model_config = {"from_attributes": True}


# =============================================================================
# Delivery Schemas
# =============================================================================

from models.delivery import SubscriptionPlan as SubscriptionPlanEnum, DeliveryStatus


class DeliveryResponse(BaseModel):
    """Schema for delivery API responses."""
    id: UUID
    user_id: UUID
    property_id: UUID
    subscription_plan: SubscriptionPlanEnum
    status: DeliveryStatus
    scheduled_for: datetime
    delivered_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    model_config = {"from_attributes": True}


class MeDeliveriesResponse(BaseModel):
    """
    Response for GET /me/deliveries endpoint.
    
    Contains the customer's next scheduled delivery (if any)
    and their delivery history (up to 20 most recent).
    """
    next_delivery: Optional[DeliveryResponse] = Field(
        default=None,
        description="Next scheduled delivery (future, not archived)"
    )
    history: list[DeliveryResponse] = Field(
        default_factory=list,
        description="Past deliveries, most recent first (limit 20)"
    )


# =============================================================================
# Florist Dashboard Schemas
# =============================================================================

from typing import Literal


class AssignedPropertyResponse(BaseModel):
    """Property info for florist's assigned properties list."""
    property_id: UUID
    property_name: str
    property_address: str


class FloristMeResponse(BaseModel):
    """
    Response for GET /florist/me endpoint.
    
    Returns florist profile with assigned properties.
    """
    florist_id: UUID
    florist_name: str
    florist_status: FloristStatus
    assigned_properties: list[AssignedPropertyResponse] = Field(
        default_factory=list,
        description="List of properties assigned to this florist"
    )


class FloristDeliveryResponse(BaseModel):
    """
    Delivery info for florist deliveries list.
    
    Includes customer and property details for display.
    """
    id: UUID
    customer_email: str
    property_id: UUID
    property_name: str
    property_address: str
    subscription_plan: SubscriptionPlanEnum
    status: DeliveryStatus
    scheduled_for: datetime


class FloristDeliveriesListResponse(BaseModel):
    """
    Response for GET /florist/deliveries endpoint.
    
    Returns list of upcoming deliveries for assigned properties.
    """
    deliveries: list[FloristDeliveryResponse] = Field(
        default_factory=list,
        description="List of scheduled deliveries for assigned properties"
    )


class UpdateDeliveryStatusRequest(BaseModel):
    """
    Request body for PATCH /florist/deliveries/{delivery_id}.
    
    Only DELIVERED or MISSED status allowed (florist actions).
    """
    status: Literal["DELIVERED", "MISSED"] = Field(
        ...,
        description="New delivery status (DELIVERED or MISSED only)"
    )

File: packages/shared/src/types/domain.ts
/**
 * Domain entity types for the Bloom platform.
 * 
 * These types match the API response structures for Property, Florist,
 * and PropertyAssignment entities.
 */

/**
 * Property status lifecycle - computed based on florist and PM assignments
 */
export const PropertyStatus = {
  CREATED: 'CREATED',           // No florist, no PM
  PENDING_FLORIST: 'PENDING_FLORIST',  // Has PM, needs florist
  PENDING_PM: 'PENDING_PM',     // Has florist, needs PM
  ACTIVE: 'ACTIVE',             // Has both florist and PM
} as const;

export type PropertyStatus = (typeof PropertyStatus)[keyof typeof PropertyStatus];

/**
 * User subscription status lifecycle
 */
export const SubscriptionStatus = {
  CREATED: 'CREATED',   // Account created, no subscription
  ACTIVE: 'ACTIVE',     // Active subscription
  PAUSED: 'PAUSED',     // Subscription paused
} as const;

export type SubscriptionStatus = (typeof SubscriptionStatus)[keyof typeof SubscriptionStatus];

/**
 * Florist status lifecycle
 */
export const FloristStatus = {
  ONBOARDING: 'ONBOARDING',
  READY: 'READY',
} as const;

export type FloristStatus = (typeof FloristStatus)[keyof typeof FloristStatus];

/**
 * Property entity - a physical location participating in Bloom's floral subscription program
 */
export interface Property {
  id: string;
  name: string;
  address: string;
  status: PropertyStatus;
  delivery_cadence: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * Enriched property entity - includes computed fields for admin table
 */
export interface EnrichedProperty {
  id: string;
  name: string;
  address: string;
  status: PropertyStatus;
  delivery_cadence: string | null;
  total_users: number;
  active_users: number;
  florist_name: string | null;
  property_manager_email: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * Florist entity - a flower vendor connected to the Bloom platform
 */
export interface Florist {
  id: string;
  name: string;
  status: FloristStatus;
  created_at: string;
}

/**
 * Property-Florist assignment - links a florist to a property for fulfillment
 */
export interface PropertyAssignment {
  id: string;
  property_id: string;
  florist_id: string;
  active: boolean;
  created_at: string;
}

// =============================================================================
// Request Types
// =============================================================================

/**
 * Request body for creating a new property
 */
export interface CreatePropertyRequest {
  name: string;
  address: string;
  delivery_cadence?: string;
}

/**
 * Request body for updating an existing property
 */
export interface UpdatePropertyRequest {
  name?: string;
  address?: string;
  status?: PropertyStatus;
  delivery_cadence?: string;
}

/**
 * Request body for creating a new florist
 */
export interface CreateFloristRequest {
  name: string;
}

/**
 * Request body for creating a property-florist assignment
 */
export interface CreatePropertyAssignmentRequest {
  property_id: string;
  florist_id: string;
}

/**
 * Request body for assigning a property manager to a property
 */
export interface AssignPMRequest {
  user_id: string;
}

// =============================================================================
// Array helpers for iteration
// =============================================================================

/**
 * Array of all property statuses for iteration
 */
export const ALL_PROPERTY_STATUSES = Object.values(PropertyStatus);

/**
 * Array of all florist statuses for iteration
 */
export const ALL_FLORIST_STATUSES = Object.values(FloristStatus);

/**
 * Array of all subscription statuses for iteration
 */
export const ALL_SUBSCRIPTION_STATUSES = Object.values(SubscriptionStatus);

// =============================================================================
// User Types
// =============================================================================

import type { UserRole } from './roles';

/**
 * User entity - a user account in the Bloom platform
 */
export interface User {
  id: string;
  email: string;
  role: UserRole;
  property_id: string | null;
  subscription_status: SubscriptionStatus | null;
  created_at: string;
}

/**
 * Admin user entity - enriched user data for admin table
 * Includes property_name resolved from property_id
 * subscription_status is null for non-CUSTOMER users
 */
export interface AdminUser {
  id: string;
  email: string;
  role: UserRole;
  property_id: string | null;
  property_name: string | null;
  subscription_status: SubscriptionStatus | null;
  created_at: string;
}

/**
 * Request body for creating a new user
 */
export interface CreateUserRequest {
  email: string;
  role: UserRole;
  password: string;
  property_id?: string;  // Only valid for PROPERTY_MANAGER role
}

/**
 * Request body for updating an existing user
 */
export interface UpdateUserRequest {
  role?: UserRole;
  property_id?: string | null;
  subscription_status?: SubscriptionStatus;
}


// =============================================================================
// Onboarding Types
// =============================================================================

/**
 * Minimal property info for public listing during onboarding
 */
export interface PropertyListItem {
  id: string;
  name: string;
  address: string;
}

/**
 * Request body for customer self-registration
 */
export interface RegisterRequest {
  email: string;
  password: string;
}

/**
 * Response from customer registration (same format as login)
 */
export interface RegisterResponse {
  access_token: string;
  token_type: string;
  user: User;
}

/**
 * Request body for customer property assignment
 */
export interface MePropertyUpdateRequest {
  property_id: string;
  unit: string;
}

/**
 * Request body for customer subscription update
 * Only ACTIVE or PAUSED allowed (not CREATED)
 */
export interface MeSubscriptionUpdateRequest {
  subscription_status: 'ACTIVE' | 'PAUSED';
}

/**
 * Request body for customer plan selection
 */
export interface MePlanUpdateRequest {
  plan: 'ESSENTIAL' | 'SIGNATURE' | 'STATEMENT';
}

// =============================================================================
// Customer Dashboard Types
// =============================================================================

/**
 * Response from GET /auth/me with enriched property data for dashboard
 * Includes property_name and property_address resolved server-side from property_id
 */
export interface MeResponse {
  id: string;
  email: string;
  role: UserRole;
  property_id: string | null;
  property_name: string | null;
  property_address: string | null;
  unit: string | null;
  subscription_status: SubscriptionStatus | null;
  subscription_plan: SubscriptionPlan | null;
  created_at: string;
}

// =============================================================================
// Delivery Types
// =============================================================================

/**
 * Subscription plan tiers
 */
export const SubscriptionPlan = {
  ESSENTIAL: 'ESSENTIAL',   // Basic floral arrangement
  SIGNATURE: 'SIGNATURE',   // Premium floral arrangement
  STATEMENT: 'STATEMENT',   // Luxury floral arrangement
} as const;

export type SubscriptionPlan = (typeof SubscriptionPlan)[keyof typeof SubscriptionPlan];

/**
 * Delivery status lifecycle
 */
export const DeliveryStatus = {
  SCHEDULED: 'SCHEDULED',   // Delivery is planned for a future date
  DELIVERED: 'DELIVERED',   // Delivery was successfully completed
  SKIPPED: 'SKIPPED',       // Customer chose to skip this delivery
  MISSED: 'MISSED',         // Delivery was not completed
} as const;

export type DeliveryStatus = (typeof DeliveryStatus)[keyof typeof DeliveryStatus];

/**
 * Array of all subscription plans for iteration
 */
export const ALL_SUBSCRIPTION_PLANS = Object.values(SubscriptionPlan);

/**
 * Array of all delivery statuses for iteration
 */
export const ALL_DELIVERY_STATUSES = Object.values(DeliveryStatus);

/**
 * Delivery entity - a scheduled or completed floral delivery
 */
export interface Delivery {
  id: string;
  user_id: string;
  property_id: string;
  subscription_plan: SubscriptionPlan;
  status: DeliveryStatus;
  scheduled_for: string;
  delivered_at: string | null;
  created_at: string;
  updated_at: string | null;
}

/**
 * Response from GET /me/deliveries
 * Contains next scheduled delivery and delivery history
 */
export interface MeDeliveriesResponse {
  next_delivery: Delivery | null;
  history: Delivery[];
}

// =============================================================================
// Florist Dashboard Types
// =============================================================================

/**
 * Assigned property info for florist dashboard
 */
export interface AssignedProperty {
  property_id: string;
  property_name: string;
  property_address: string;
}

/**
 * Response from GET /florist/me
 * Returns florist profile with assigned properties
 */
export interface FloristMeResponse {
  florist_id: string;
  florist_name: string;
  florist_status: FloristStatus;
  assigned_properties: AssignedProperty[];
}

/**
 * Delivery info for florist deliveries list
 * Includes customer and property details for display
 */
export interface FloristDelivery {
  id: string;
  customer_email: string;
  property_id: string;
  property_name: string;
  property_address: string;
  subscription_plan: SubscriptionPlan;
  status: DeliveryStatus;
  scheduled_for: string;
}

/**
 * Response from GET /florist/deliveries
 * Returns list of upcoming deliveries for assigned properties
 */
export interface FloristDeliveriesListResponse {
  deliveries: FloristDelivery[];
}

/**
 * Request body for updating delivery status
 * Only DELIVERED or MISSED allowed (florist actions)
 */
export interface UpdateDeliveryStatusRequest {
  status: 'DELIVERED' | 'MISSED';
}
